plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java

```diff
@@ -5611,4 +5612,122 @@ public boolean attachConfigDriveToMigratedVm(Connection conn, String vmName, Str
 
     }
 
+    /**
+     * Get Diagnostics Data API
+     * Copy zip file from system vm and copy file directly to secondary storage
+     *
+     * @param conn
+     * @param cmd
+     * @return
+     */
+    public Answer copyDiagnosticsFileToSecondaryStorage(Connection conn, CopyToSecondaryStorageCommand cmd) {
+        String secondaryStorageUrl = cmd.getSecondaryStorageUrl();
+        String vmIP = cmd.getSystemVmIp();
+        //fileName = /root/<fileName>
+        String fileName = cmd.getFileName();
+
+        String localDir = null;
+        boolean success;
+
+        // Mount Secondary storage
+        String secondaryStorageMountPath = null;
+        try {
+            URI uri = new URI(secondaryStorageUrl);
+            secondaryStorageMountPath = uri.getHost() + ":" + uri.getPath();
+            localDir = "/var/cloud_mount/" + UUID.nameUUIDFromBytes(secondaryStorageMountPath.getBytes());
+            mountNfs(conn, secondaryStorageMountPath, localDir);
+            String diagnosticsFileDir = "diagnostics_data";
+
+            String secStoragePath = localDir + "/" + diagnosticsFileDir;
+            Pair<Boolean, String> copyDetails = copyFilesToSecondaryStorage(vmIP, fileName, secStoragePath);
+            success = copyDetails.first();
+
+            if (!success) s_logger.error("failed to copy zip file to sec storage: " + copyDetails.second());
+
+        } catch (URISyntaxException e) {
+            throw new CloudRuntimeException("URI exception caught during decoding of secondary storage URI " + secondaryStorageUrl, e);
+        } finally {
+            if (localDir != null) umountNfs(conn, secondaryStorageMountPath, localDir);
+            // Cleanup file in system VM
+            cleanUpFileInVm(vmIP, fileName);
+        }
+        return new Answer(cmd, success, "");
+    }
+
+    /**
+     * Get Diagnostics Data API
+     * Copy zip file from system vm directly to diagnostics folder in secondary storage
+     * Secondary storage needs to be mounted on hypervisor host at this point
+     *
+     * @param vmIP           vm ssh IP address
+     * @param fileName       /root/diagnostics_files_xxxx.tar
+     * @param secStoragePath the path to 'diagnostics_data' directory in mounted secondary storage e.x /var/cloud_mout/uuid/diagnostics_data
+     *                       directory is created if it does not exist
+     * @return
+     */
+    protected Pair<Boolean, String> copyFilesToSecondaryStorage(String vmIP, String fileName, String secStoragePath) {
+        final String DEFAULT_PORT = "3922";
```

Comment:
final strings in a method does not seems like a sound way to go to me.
I'd suggest global setting (or, if you hate me, at least class level constants)

Reply:
i take that back, use global settings.

Reply:
Moved to class level constant, for as long as we use port 3922 for secure mgmt-agent communication, I do not think global settings are necessary. At least for this feature